File: web-app/tests/LocationServiceParity.test.ts
/* eslint-env browser */
/* global GeolocationPosition */
import { describe, it, expect, vi, afterEach } from 'vitest';
import {
  LocationService,
  setPositionGetter,
  setIsoCodeGetter
} from '../src/services/LocationService';
import { ApiQuotaLedger } from '../src/utils/ApiQuotaLedger';
import type { CountrySetting } from '../src/repositories/CountrySettingRepository';

class FakeRepo {
  saved?: CountrySetting;
  async save(s: CountrySetting) {
    this.saved = s;
  }
  async load() {
    return null;
  }
}

afterEach(() => {
  vi.useRealTimers();
  vi.restoreAllMocks();
  setPositionGetter(() =>
    new Promise(resolve =>
      resolve({ coords: { latitude: 0, longitude: 0 } } as GeolocationPosition)
    )
  );
  setIsoCodeGetter(async () => null);
});

describe('LocationService parity with Dart', () => {
  it('increments ledger after window expiry', async () => {
    vi.useFakeTimers();
    const repo = new FakeRepo();
    const ledger = new ApiQuotaLedger(1, 50);
    const svc = new LocationService(repo as any, ledger);
    setPositionGetter(
      vi.fn().mockResolvedValue({ coords: { latitude: 1, longitude: 2 } }) as any
    );
    setIsoCodeGetter(vi.fn().mockResolvedValue('US') as any);

    await svc.resolveCountry();
    expect(ledger.isSafe()).toBe(false);
    vi.advanceTimersByTime(60);
    expect(ledger.isSafe()).toBe(true);
    await svc.resolveCountry();
    expect(repo.saved?.iso2).toBe('US');
  });

  it('throws when ledger disallows', async () => {
    const repo = new FakeRepo();
    const ledger = new ApiQuotaLedger(0);
    const svc = new LocationService(repo as any, ledger);
    setPositionGetter(
      vi.fn().mockResolvedValue({ coords: { latitude: 1, longitude: 2 } }) as any
    );
    setIsoCodeGetter(vi.fn().mockResolvedValue('US') as any);
    await expect(svc.resolveCountry()).rejects.toThrow('quota');
  });

  it('throws when iso code missing', async () => {
    const repo = new FakeRepo();
    const ledger = new ApiQuotaLedger(1);
    const svc = new LocationService(repo as any, ledger);
    setPositionGetter(
      vi.fn().mockResolvedValue({ coords: { latitude: 1, longitude: 2 } }) as any
    );
    setIsoCodeGetter(vi.fn().mockResolvedValue(null) as any);
    await expect(svc.resolveCountry()).rejects.toThrow();
  });
});

File: mobile-app/packages/services/test/location_service_parity_test.dart
import 'package:smwa_services/services.dart';
import 'package:smwa_services/src/location_service.dart' as loc;
import 'package:smwa_services/src/country_setting_repository.dart';
import 'package:smwa_services/src/country_setting.dart';
import 'package:geolocator/geolocator.dart';
import 'package:flutter_test/flutter_test.dart';

class _FakeRepo implements CountrySettingRepository {
  CountrySetting? saved;

  @override
  Future<CountrySetting?> load() async => null;

  @override
  Future<void> save(CountrySetting setting) async {
    saved = setting;
  }
}

class _FakeGeolocator extends GeolocatorPlatform {
  @override
  Future<LocationPermission> checkPermission() async =>
      LocationPermission.always;

  @override
  Future<LocationPermission> requestPermission() async =>
      LocationPermission.always;

  @override
  Future<bool> isLocationServiceEnabled() async => true;
}

Position _pos() => Position(
      latitude: 1.0,
      longitude: 2.0,
      timestamp: DateTime.now(),
      accuracy: 0,
      altitude: 0,
      altitudeAccuracy: 0,
      heading: 0,
      headingAccuracy: 0,
      speed: 0,
      speedAccuracy: 0,
    );

void main() {
  TestWidgetsFlutterBinding.ensureInitialized();
  GeolocatorPlatform.instance = _FakeGeolocator();
  tearDown(() {
    loc.positionGetter = () => Geolocator.getCurrentPosition();
    loc.isoCodeGetter = (lat, lon) async => null;
  });

  test('ledger increments again after window expiry', () async {
    loc.positionGetter = () async => _pos();
    loc.isoCodeGetter = (lat, lon) async => 'US';
    final repo = _FakeRepo();
    final ledger = ApiQuotaLedger(1, const Duration(milliseconds: 50));
    final svc = LocationService(repository: repo, ledger: ledger);

    await svc.resolveCountry();
    expect(ledger.isSafe(), isFalse);
    await Future.delayed(const Duration(milliseconds: 60));
    expect(ledger.isSafe(), isTrue);
    await svc.resolveCountry();
    expect(repo.saved?.iso2, 'US');
  });

  test('throws when ledger disallows', () async {
    loc.positionGetter = () async => _pos();
    loc.isoCodeGetter = (lat, lon) async => 'US';
    final svc =
        LocationService(repository: _FakeRepo(), ledger: ApiQuotaLedger(0));
    expect(() async => await svc.resolveCountry(), throwsException);
  });

  test('throws when iso not found', () async {
    loc.positionGetter = () async => _pos();
    loc.isoCodeGetter = (lat, lon) async => null;
    final svc =
        LocationService(repository: _FakeRepo(), ledger: ApiQuotaLedger(1));
    expect(() async => await svc.resolveCountry(), throwsException);
  });
}
